dlc-manager/src/sub_channel_manager.rs

```diff
@@ -1079,50 +1110,139 @@ where
-                warn!("Got close notification while in ClosedPunished.");
-                return Ok(());
-            }
-            SubChannelState::Rejected => {
-                info!("Counterparty closed channel in rejected state, marking as counter closed");
-                sub_channel.state = SubChannelState::CounterOnChainClosed;
-                (None, None)
+        let (updated_channel, updated_contract) = match closure_reason {
+            ClosureReason::CounterpartyForceClosed { .. }
+            | ClosureReason::CommitmentTxConfirmed => match sub_channel.state {
+                SubChannelState::Offered(_)
+                | SubChannelState::Accepted(_)
+                | SubChannelState::Confirmed(_)
+                | SubChannelState::CloseAccepted(_)
+                | SubChannelState::CloseConfirmed(_)
+                | SubChannelState::Finalized(_) => {
+                    let dlc_channel_id = sub_channel
+                        .get_dlc_channel_id(0)
+                        .expect("to have a channel id");
+                    let (closed_channel, closed_contract) =
+                        self.get_closed_dlc_channel_and_contract(dlc_channel_id, true)?;
+                    sub_channel.state = SubChannelState::CounterOnChainClosed;
+                    chain_monitor.cleanup_channel(sub_channel.channel_id);
+                    chain_monitor.cleanup_channel(dlc_channel_id);
+                    (Some(closed_channel), Some(closed_contract))
+                }
+                SubChannelState::OffChainClosed => {
+                    chain_monitor.cleanup_channel(sub_channel.channel_id);
+                    sub_channel.state = SubChannelState::CounterOnChainClosed;
+                    (None, None)
+                }
+                SubChannelState::Signed(_)
+                | SubChannelState::Closing(_)
+                | SubChannelState::CloseOffered(_) => {
+                    error!("Got notification of LN channel closure by counter party in a state where we do not expect it.");
+                    return Ok(());
+                }
+                SubChannelState::OnChainClosed | SubChannelState::CounterOnChainClosed => {
+                    info!(
+                        "Channel close notification received for channel: {:?}",
+                        sub_channel.channel_id
+                    );
+                    return Ok(());
+                }
+                SubChannelState::ClosedPunished(_) => {
+                    warn!("Got close notification while in ClosedPunished.");
+                    return Ok(());
+                }
+                SubChannelState::Rejected => {
+                    info!(
+                        "Counterparty closed channel in rejected state, marking as counter closed"
+                    );
+                    sub_channel.state = SubChannelState::CounterOnChainClosed;
+                    (None, None)
+                }
```

Comment:
I reckon we could move this one to the branch with `OffChainClosed`. They're kind of equivalent: there is no subchannel.

Reply:
Yeah but in the `OffChainClosed` state you need to cleanup the chain monitor, would be a bit weird to do it anyway for rejected channels I feel.

Reply:
Ah, fair. Didn't notice the difference!